File: app/src/main/java/tv/spacedentist/android/chromecast/SDChromecastManager.java
package tv.spacedentist.android.chromecast;

import com.google.android.gms.cast.CastDevice;
import com.google.android.gms.cast.framework.CastContext;
import com.google.android.gms.cast.framework.CastSession;
import com.google.android.gms.cast.framework.CastStateListener;
import com.google.android.gms.cast.framework.SessionManagerListener;
import com.google.android.gms.common.api.Status;

import java.util.HashSet;
import java.util.Set;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import tv.spacedentist.android.BuildConfig;
import tv.spacedentist.android.SDComponent;
import tv.spacedentist.android.util.SDLogger;

/**
 * This is where most of the Chormecast logic happens
 */
public class SDChromecastManager implements CastStateListener {

    private static final String TAG = "SDChromecastManager";

    private final Set<CastStateListener> mListeners = new HashSet<>();
    private final SessionManagerListener<CastSession> mSessionManagerListener;

    @Inject SDLogger mLogger;
    @Inject CastContext mCastContext;

    public SDChromecastManager(SDComponent component) {
        component.inject(this);

        mSessionManagerListener = new SDSessionManagerListener(mLogger);
    }

    public void addCastStateListener(CastStateListener listener) {
        mListeners.add(listener);
    }

    public void removeCastStateListener(CastStateListener listener) {
        mListeners.remove(listener);
    }

    @Override
    public void onCastStateChanged(int state) {
        mLogger.i(TAG, "onCastStateChanged: " + state);
        for (CastStateListener listener : mListeners) {
            listener.onCastStateChanged(state);
        }
    }

    public @Nullable String getSelectedDeviceFriendlyName() {
        return mCastContext.getSessionManager().getCurrentCastSession().getCastDevice().getFriendlyName();
    }

    private void onMessageResult(Status status) {
        if (!status.isSuccess()) {
            mLogger.e(TAG, "Sending message failed");
        }
    };

    public void sendChromecastMessage(String message) {
        mLogger.i(TAG, "sending message: '" + message + "' state: " + mCastContext.getCastState());

        CastSession castSession = mCastContext.getSessionManager().getCurrentCastSession();

        if (castSession != null) {
            mLogger.i(TAG, "there's a session");
            castSession
                    .sendMessage(BuildConfig.CHROMECAST_APP_NAMESPACE, message)
                    .setResultCallback(this::onMessageResult);
        } else {
            mLogger.i(TAG, "no current session");
        }
    }

    public void onResume() {
        mCastContext.getSessionManager().addSessionManagerListener(mSessionManagerListener, CastSession.class);
        mCastContext.addCastStateListener(this);
    }

    public void onPause() {
        mCastContext.getSessionManager().addSessionManagerListener(mSessionManagerListener, CastSession.class);
        mCastContext.removeCastStateListener(this);
    }

    public void endCurrentSession() {
        mCastContext.getSessionManager().endCurrentSession(true);
    }

    public CastDevice getCastDevice() {
        return mCastContext.getSessionManager().getCurrentCastSession().getCastDevice();
    }

    public int getCurrentCastState() {
        return mCastContext.getCastState();
    }
}

File: app/src/main/java/tv/spacedentist/android/chromecast/SDChromecastManager.kt
package tv.spacedentist.android.chromecast

import com.google.android.gms.cast.CastDevice
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.CastSession
import com.google.android.gms.cast.framework.CastStateListener
import com.google.android.gms.cast.framework.SessionManagerListener
import com.google.android.gms.common.api.Status
import tv.spacedentist.android.BuildConfig
import tv.spacedentist.android.SDComponent
import tv.spacedentist.android.util.SDLogger
import java.util.*

/**
 * This used to be where most of the Chormecast logic happens, but now it mostly just redistributes
 * cast state changed callbacks
 */
class SDChromecastManager(private val component: SDComponent) : CastStateListener {

    private val mListeners = HashSet<CastStateListener>()
    private val mSessionManagerListener: SessionManagerListener<CastSession>

    private val logger: SDLogger
        get() = component.logger

    private val castContext: CastContext
        get() = component.castContext

    val selectedDeviceFriendlyName: String?
        get() = castContext.sessionManager.currentCastSession?.castDevice?.friendlyName

    val castDevice: CastDevice?
        get() = castContext.sessionManager.currentCastSession?.castDevice

    val currentCastState: Int
        get() = castContext.castState

    init {
        mSessionManagerListener = SDSessionManagerListener(logger)
    }

    fun addCastStateListener(listener: CastStateListener) {
        mListeners.add(listener)
    }

    fun removeCastStateListener(listener: CastStateListener) {
        mListeners.remove(listener)
    }

    override fun onCastStateChanged(state: Int) {
        logger.i(TAG, "onCastStateChanged: $state")
        for (listener in mListeners) {
            listener.onCastStateChanged(state)
        }
    }

    private fun onMessageResult(status: Status) {
        if (!status.isSuccess) {
            logger.e(TAG, "Sending message failed")
        }
    }

    fun sendChromecastMessage(message: String) {
        logger.i(TAG, "sending message: '" + message + "' state: " + castContext.castState)

        castContext.sessionManager.currentCastSession?.let {
            it.sendMessage(BuildConfig.CHROMECAST_APP_NAMESPACE, message)
                    .setResultCallback(this::onMessageResult)
        }
    }

    fun onResume() {
        castContext.sessionManager.addSessionManagerListener(mSessionManagerListener, CastSession::class.java)
        castContext.addCastStateListener(this)
    }

    fun onPause() {
        castContext.sessionManager.addSessionManagerListener(mSessionManagerListener, CastSession::class.java)
        castContext.removeCastStateListener(this)
    }

    fun endCurrentSession() {
        castContext.sessionManager.endCurrentSession(true)
    }

    companion object {
        private const val TAG = "SDChromecastManager"
    }
}
